presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTreeTranslator.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.expressions.translator;
+
+import com.facebook.presto.spi.function.FunctionMetadataManager;
+import com.facebook.presto.spi.function.StandardFunctionResolution;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.DeterminismEvaluator;
+import com.facebook.presto.spi.relation.InputReferenceExpression;
+import com.facebook.presto.spi.relation.LambdaDefinitionExpression;
+import com.facebook.presto.spi.relation.LogicalRowExpressions;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.RowExpressionVisitor;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+
+public class RowExpressionTreeTranslator<T, C>
+{
+    private final RowExpressionTranslator<T, C> rowExpressionTranslator;
+    private final RowExpressionVisitor<TranslatedExpression<T>, C> visitor;
+
+    private RowExpressionTreeTranslator(RowExpressionTranslator<T, C> rowExpressionTranslator)
+    {
+        this.rowExpressionTranslator = rowExpressionTranslator;
+        this.visitor = new TranslatingVisitor();
+    }
+
+    public RowExpressionVisitor<TranslatedExpression<T>, C> getVisitor()
+    {
+        return visitor;
+    }
+
+    public static <T, C> TranslatedExpression<T> translateWith(
+            RowExpression expression,
+            DeterminismEvaluator determinismEvaluator,
+            StandardFunctionResolution functionResolution,
+            FunctionMetadataManager functionMetadataManager,
+            RowExpressionTranslator<T, C> translator,
+            C context)
+    {
+        LogicalRowExpressions logicalRowExpressions = new LogicalRowExpressions(determinismEvaluator, functionResolution, functionMetadataManager);
```

Comment:
We don't need to put this logic in the `RowExpressionTreeTranslator`, the user of `translateWith` need to do before they call `translateWith`.

Reply:
Are there any side-effects of this assumption? I.e. if the user does not convert the expression/predicate to conjunctive form, could their be any resultant issues other than building the *minimal* converted rowexpression? I just want to make sure that by moving this out we dont make the code less fool-proof.

Reply:
I don't think this `convertToConjunctiveNormalForm` is needed for all translations, it is only used to make predicate pushdown easier, so definitely remove it from the common interface.